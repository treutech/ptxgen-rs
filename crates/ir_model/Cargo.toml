
[package]
name = "ir-model"
version = "0.1.0"
edition = "2024"
license = "Apache-2.0"
authors = ["Raul Estrada <restrada@treutech.io>"]
homepage = "https://treutech.io/ptxgen-rs"
repository = "https://github.com/treutech/ptxgen-rs"

[dependencies]
serde = { version = "1.0.219", features = ["derive"] }
