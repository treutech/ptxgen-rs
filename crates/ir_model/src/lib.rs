// Copyright 2025 Raul Estrada <restrada@treutech.io>
// SPDX-License-Identifier: Apache-2.0
//
// This file is part of the PTXGEN-RS project by Treu Technologies.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use serde::Serialize;

#[derive(Debug, Serialize)]
pub enum Instruction {
    Load {
        function: String,
        dst: String,
        src: String,
    },
    Store {
        function: String,
        dst: String,
        value: String,
    },
    Add {
        function: String,
        dst: String,
        lhs: String,
        rhs: String,
    },
    FAdd {
        function: String,
        dst: String,
        lhs: String,
        rhs: String,
    },
    FMul {
        function: String,
        dst: String,
        lhs: String,
        rhs: String,
    },
    ICmp {
        function: String,
        dst: String,
        lhs: String,
        rhs: String,
        op: String,
    },
    GetElementPtr {
        function: String,
        dst: String,
        base: String,
        index: String,
    },
    Alloca {
        function: String,
        dst: String,
        ty: String,
        align: u32,
    },
    Br {
        function: String,
        cond: Option<String>,
        target_true: String,
        target_false: Option<String>,
    },
    Ret {
        function: String,
    },
    Unhandled {
        function: String,
        text: String,
    },
}

impl Instruction {
    pub fn function_name(&self) -> &str {
        match self {
            Instruction::Load { function, .. }
            | Instruction::Store { function, .. }
            | Instruction::Add { function, .. }
            | Instruction::FAdd { function, .. }
            | Instruction::FMul { function, .. }
            | Instruction::ICmp { function, .. }
            | Instruction::GetElementPtr { function, .. }
            | Instruction::Alloca { function, .. }
            | Instruction::Br { function, .. }
            | Instruction::Ret { function }
            | Instruction::Unhandled { function, .. } => function,
        }
    }

    pub fn used_operands(&self) -> Vec<&str> {
        use Instruction::*;
        match self {
            FMul { dst, lhs, rhs, .. }
            | FAdd { dst, lhs, rhs, .. }
            | Add { dst, lhs, rhs, .. }
            | ICmp { dst, lhs, rhs, .. } => vec![dst, lhs, rhs],

            Load { dst, src, .. } => vec![dst, src],
            Store { dst, value, .. } => vec![dst, value],

            Alloca { dst, .. } => vec![dst],
            GetElementPtr {
                dst, base, index, ..
            } => vec![dst, base, index],

            Unhandled { text, .. } => vec![text],
            _ => vec![],
        }
    }
}
