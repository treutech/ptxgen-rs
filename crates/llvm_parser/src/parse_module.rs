use anyhow::{Context, Result};
use llvm_ir::Module;
use std::fs;
use std::path::Path;

/// Parse an LLVM IR (.ll) file into an `llvm_ir::Module`
pub fn parse_module<P: AsRef<Path>>(path: P) -> Result<Module> {
    let ll_text = fs::read_to_string(&path)
        .with_context(|| format!("Failed to read file: {}", path.as_ref().display()))?;

    Module::from_ir_str(&ll_text)
        .map_err(|e| anyhow::anyhow!(e))
        .with_context(|| format!("Failed to parse LLVM IR in file: {}", path.as_ref().display()))
}
