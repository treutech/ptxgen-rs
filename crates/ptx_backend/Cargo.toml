
[package]
name = "ptx-backend"
version = "0.1.0"
edition = "2024"
license = "Apache-2.0"
authors = ["Raul Estrada <restrada@treutech.io>"]
homepage = "https://treutech.io/ptxgen-rs"
repository = "https://github.com/treutech/ptxgen-rs"

[dependencies]
ir-model = { path = "../ir_model" }
llvm-parser = { path = "../llvm_parser" }
clap = { version = "4.5.41", features = ["derive"] }
assert_cmd = "2.0.17"
anyhow = "1.0.98"
llvm-ir = "0.11.3"

[[bin]]
name = "ptx-backend"
path = "src/main.rs"

[dev-dependencies]
insta = "1.43.1"
