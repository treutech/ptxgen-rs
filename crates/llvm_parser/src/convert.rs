// Copyright 2025 Raul Estrada <restrada@treutech.io>
// SPDX-License-Identifier: Apache-2.0
//
// This file is part of the PTXGEN-RS project by Treu Technologies.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use ir_model::Instruction;
use llvm_ir::instruction::Instruction as LlvmInst;

pub fn lower(function: &str, instr: &LlvmInst) -> Instruction {
    use LlvmInst::*;
    match instr {
        FMul(f) => Instruction::FMul {
            function: function.to_string(),
            dst: f.dest.to_string(),
            lhs: f.operand0.to_string(),
            rhs: f.operand1.to_string(),
        },
        FAdd(f) => Instruction::FAdd {
            function: function.to_string(),
            dst: f.dest.to_string(),
            lhs: f.operand0.to_string(),
            rhs: f.operand1.to_string(),
        },
        Load(l) => Instruction::Load {
            function: function.to_string(),
            dst: l.dest.to_string(),
            src: l.address.to_string(),
        },
        Store(s) => Instruction::Store {
            function: function.to_string(),
            dst: s.address.to_string(),
            value: s.value.to_string(),
        },
        Alloca(a) => Instruction::Alloca {
            function: function.to_string(),
            dst: a.dest.to_string(),
            ty: format!("{:?}", a.allocated_type),
            align: a.alignment,
        },
        ICmp(cmp) => Instruction::ICmp {
            function: function.to_string(),
            dst: cmp.dest.to_string(),
            op: format!("{:?}", cmp.predicate),
            lhs: cmp.operand0.to_string(),
            rhs: cmp.operand1.to_string(),
        },
        FCmp(cmp) => Instruction::FCmp {
            function: function.to_string(),
            dst: cmp.dest.to_string(),
            op: format!("{:?}", cmp.predicate),
            lhs: cmp.operand0.to_string(),
            rhs: cmp.operand1.to_string(),
        },
        Add(add) => Instruction::Add {
            function: function.to_string(),
            dst: add.dest.to_string(),
            lhs: add.operand0.to_string(),
            rhs: add.operand1.to_string(),
        },
        Sub(s) => Instruction::Sub {
            function: function.to_string(),
            dst: s.dest.to_string(),
            lhs: s.operand0.to_string(),
            rhs: s.operand1.to_string(),
        },
        FSub(s) => Instruction::FSub {
            function: function.to_string(),
            dst: s.dest.to_string(),
            lhs: s.operand0.to_string(),
            rhs: s.operand1.to_string(),
        },
        Mul(m) => Instruction::Mul {
            function: function.to_string(),
            dst: m.dest.to_string(),
            lhs: m.operand0.to_string(),
            rhs: m.operand1.to_string(),
        },
        UDiv(d) => Instruction::UDiv {
            function: function.to_string(),
            dst: d.dest.to_string(),
            lhs: d.operand0.to_string(),
            rhs: d.operand1.to_string(),
        },
        SDiv(d) => Instruction::SDiv {
            function: function.to_string(),
            dst: d.dest.to_string(),
            lhs: d.operand0.to_string(),
            rhs: d.operand1.to_string(),
        },
        URem(r) => Instruction::URem {
            function: function.to_string(),
            dst: r.dest.to_string(),
            lhs: r.operand0.to_string(),
            rhs: r.operand1.to_string(),
        },
        SRem(r) => Instruction::SRem {
            function: function.to_string(),
            dst: r.dest.to_string(),
            lhs: r.operand0.to_string(),
            rhs: r.operand1.to_string(),
        },
        FDiv(d) => Instruction::FDiv {
            function: function.to_string(),
            dst: d.dest.to_string(),
            lhs: d.operand0.to_string(),
            rhs: d.operand1.to_string(),
        },
        FRem(r) => Instruction::FRem {
            function: function.to_string(),
            dst: r.dest.to_string(),
            lhs: r.operand0.to_string(),
            rhs: r.operand1.to_string(),
        },
        GetElementPtr(gep) => Instruction::GetElementPtr {
            function: function.to_string(),
            dst: gep.dest.to_string(),
            base: gep.address.to_string(),
            index: gep
                .indices
                .iter()
                .map(|i| i.to_string())
                .collect::<Vec<_>>()
                .join(", "),
        },
        Phi(p) => Instruction::Phi {
            function: function.to_string(),
            dst: p.dest.to_string(),
            incoming: p
                .incoming_values
                .iter()
                .map(|(val, label)| {
                    (
                        match label {
                            llvm_ir::Name::Name(s) => s.as_str().to_string(),
                            llvm_ir::Name::Number(n) => format!("{}", n),
                        },
                        val.to_string(),
                    )
                })
                .collect(),
        },
        Select(sel) => Instruction::Select {
            function: function.to_string(),
            dst: sel.dest.to_string(),
            cond: sel.condition.to_string(),
            val_true: sel.true_value.to_string(),
            val_false: sel.false_value.to_string(),
        },
        BitCast(bc) => Instruction::Bitcast {
            function: function.to_string(),
            dst: bc.dest.to_string(),
            src: bc.operand.to_string(),
        },
        ZExt(z) => Instruction::ZExt {
            function: function.to_string(),
            dst: z.dest.to_string(),
            src: z.operand.to_string(),
        },
        Trunc(t) => Instruction::Trunc {
            function: function.to_string(),
            dst: t.dest.to_string(),
            src: t.operand.to_string(),
        },
        Call(c) => {
            let target = match &c.function {
                either::Either::Right(llvm_ir::Operand::ConstantOperand(const_ref)) => {
                    match const_ref.as_ref() {
                        llvm_ir::constant::Constant::GlobalReference { name, .. } => match name {
                            llvm_ir::Name::Name(s) => s.to_string(),
                            llvm_ir::Name::Number(n) => format!("{}", n),
                        },
                        _ => "unknown_fn".to_string(),
                    }
                }
                _ => "unknown_fn".to_string(),
            };

            Instruction::Call {
                function: function.to_string(),
                target,
                args: c.arguments.iter().map(|a| a.0.to_string()).collect(),
            }
        }
        _ => Instruction::Unhandled {
            function: function.to_string(),
            text: format!("{:?}", instr),
        },
    }
}

use llvm_ir::Terminator;

pub fn lower_terminator(func: &str, term: &Terminator) -> Instruction {
    match term {
        Terminator::Ret(_) => Instruction::Ret {
            function: func.to_string(),
        },
        Terminator::Br(br) => Instruction::Br {
            function: func.to_string(),
            cond: None,
            target_true: br.dest.to_string(),
            target_false: None,
        },
        Terminator::CondBr(br) => Instruction::Br {
            function: func.to_string(),
            cond: Some(format!("{}", br.condition)),
            target_true: br.true_dest.to_string(),
            target_false: Some(br.false_dest.to_string()),
        },
        _ => Instruction::Unhandled {
            function: func.to_string(),
            text: format!("{:?}", term),
        },
    }
}
