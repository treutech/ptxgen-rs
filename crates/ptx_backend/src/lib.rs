// Copyright 2025 Raul Estrada <restrada@treutech.io>
// SPDX-License-Identifier: Apache-2.0
//
// This file is part of the PTXGEN-RS project by Treu Technologies.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

mod utils;
use crate::utils::{clean_operand, get_register_type};
use ir_model::Instruction;
use std::collections::HashMap;

pub fn lower_function(
    name: &str,
    all_instrs: &[(String, Vec<Instruction>)],
    target: &str,
) -> Vec<String> {
    let mut output = vec![];

    output.push(format!("// Function: {}", name));
    output.push(emit_header(target));
    output.push(format!(".entry {} {{", clean_operand(name)));

    let flat_instrs: Vec<&Instruction> = all_instrs
        .iter()
        .flat_map(|(_, instrs)| instrs.iter())
        .collect();

    output.push(declare_registers(&flat_instrs));

    for (block_name, instrs) in all_instrs {
        if instrs.is_empty() {
            continue;
        }
        output.push(format!("{}:", clean_operand(block_name)));
        for instr in instrs {
            let line = to_ptx(instr, &flat_instrs);
            output.push(format!("    {}", line));
        }
    }

    let last_instr = flat_instrs.iter().rev().find(|i| {
        !matches!(
            i,
            Instruction::Alloca { .. } | Instruction::GetElementPtr { .. }
        )
    });

    if let Some(instr) = last_instr {
        match instr {
            Instruction::Ret { .. } | Instruction::Br { .. } => {}
            _ => output.push("    ret;".into()),
        }
    } else {
        output.push("    ret;".into());
    }

    output.push("}".into());
    output
}

fn emit_header(target: &str) -> String {
    format!(".version 7.0\n.target {}\n.address_size 64\n", target)
}

pub fn declare_registers(instrs: &[&Instruction]) -> String {
    let mut reg_types: HashMap<String, &str> = HashMap::new();

    for reg in ["x", "y", "out"] {
        reg_types.entry(reg.to_string()).or_insert("s32");
    }

    for instr in instrs {
        for name in instr.used_operands() {
            let reg = clean_operand(&name);
            if let Some(reg_type) = get_register_type(instr, &name) {
                match reg_types.get(&reg) {
                    Some(&existing_type) if existing_type != reg_type => {
                        reg_types.insert(reg.clone(), dominant_type(existing_type, reg_type));
                    }
                    None => {
                        reg_types.insert(reg.clone(), reg_type);
                    }
                    _ => {}
                }
            }
        }
    }

    // üëá ESTE BLOQUE DEBE VENIR ANTES DEL USO POSTERIOR DE reg_types
    let mut temp_regs: Vec<String> = vec![];
    for instr in instrs {
        if let Instruction::GetElementPtr { dst, .. } = instr {
            let dst = clean_operand(dst);
            temp_regs.push(dst.clone()); // e.g., x_
            temp_regs.push(format!("{}_offset", dst)); // e.g., x__offset
        }
    }
    for reg in temp_regs {
        reg_types.entry(reg).or_insert("s32");
    }

    let mut f32_regs = vec![];
    let mut s32_regs = vec![];
    let mut pred_regs = vec![];

    for (reg, ty) in reg_types {
        match ty {
            "f32" => f32_regs.push(reg),
            "s32" => s32_regs.push(reg),
            "pred" => pred_regs.push(reg),
            _ => {}
        }
    }

    f32_regs.sort();
    s32_regs.sort();
    pred_regs.sort();

    // Detectar variables alocadas localmente
    let mut local_vars = vec![];
    for instr in instrs {
        if let Instruction::Alloca { dst, ty, .. } = instr {
            let reg = clean_operand(dst);
            let ty_str = if ty.contains("f32") { "f32" } else { "s32" };
            local_vars.push((reg, ty_str));
        }
    }
    local_vars.sort();
    let mut out = vec![];

    for (reg, ty) in local_vars {
        out.push(format!(".local .{} {};", ty, reg));
    }

    if !f32_regs.is_empty() {
        out.push(format!(
            ".reg .f32 {};",
            f32_regs
                .iter()
                .map(|r| format!("%{}", r))
                .collect::<Vec<_>>()
                .join(", ")
        ));
    }
    if !s32_regs.is_empty() {
        out.push(format!(
            ".reg .s32 {};",
            s32_regs
                .iter()
                .map(|r| format!("%{}", r))
                .collect::<Vec<_>>()
                .join(", ")
        ));
    }
    if !pred_regs.is_empty() {
        out.push(format!(
            ".reg .pred {};",
            pred_regs
                .iter()
                .map(|r| format!("%{}", r))
                .collect::<Vec<_>>()
                .join(", ")
        ));
    }

    // ‚ö†Ô∏è Parche urgente para declarar manualmente registros faltantes:
    s32_regs.push("i".to_string());
    s32_regs.push("sum".to_string());
    s32_regs.push("loop".to_string());
    s32_regs.push("body".to_string());
    s32_regs.push("exit".to_string());

    out.push(String::new());

    out.join("\n")
}

fn dominant_type<'a>(a: &'a str, b: &'a str) -> &'a str {
    if a == "s32" || b == "s32" {
        "s32"
    } else if a == "f32" || b == "f32" {
        "f32"
    } else {
        "pred"
    }
}

fn register_type(reg_name: &str, instrs: &[&Instruction]) -> &'static str {
    for instr in instrs {
        if let Some(t) = get_register_type(instr, reg_name) {
            return t;
        }
    }
    "s32" // default fallback
}

pub fn to_ptx(instr: &Instruction, all_instrs: &[&Instruction]) -> String {
    use Instruction::*;

    fn reg(op: &str) -> String {
        let clean = clean_operand(op);
        if clean.starts_with('%') {
            clean
        } else {
            format!("%{}", clean)
        }
    }

    fn mem(op: &str) -> String {
        let clean = clean_operand(op);
        if clean.starts_with('%') {
            format!("[{}]", clean)
        } else {
            format!("[%{}]", clean)
        }
    }

    fn is_local(name: &str, instrs: &[&Instruction]) -> bool {
        instrs
            .iter()
            .any(|i| matches!(i, Alloca { dst, .. } if dst == name))
    }

    match instr {
        FMul { dst, lhs, rhs, .. } => {
            format!("mul.f32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        FAdd { dst, lhs, rhs, .. } => {
            format!("add.f32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        Add { dst, lhs, rhs, .. } => {
            format!("add.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        Sub { dst, lhs, rhs, .. } => {
            format!("sub.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        FSub { dst, lhs, rhs, .. } => {
            format!("sub.f32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        Mul { dst, lhs, rhs, .. } => {
            format!("mul.lo.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        UDiv { dst, lhs, rhs, .. } => {
            format!("div.u32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        SDiv { dst, lhs, rhs, .. } => {
            format!("div.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        URem { dst, lhs, rhs, .. } => {
            format!("rem.u32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        SRem { dst, lhs, rhs, .. } => {
            format!("rem.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        FDiv { dst, lhs, rhs, .. } => {
            format!("div.f32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        FRem { dst, lhs, rhs, .. } => {
            format!("rem.f32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        ICmp { dst, lhs, rhs, .. } => {
            format!("setp.lt.s32 {}, {}, {};", reg(dst), reg(lhs), reg(rhs))
        }
        Load { dst, src, .. } => {
            let ty = register_type(dst, all_instrs);
            let space = if is_local(src, all_instrs) {
                "local"
            } else {
                "global"
            };
            format!("ld.{space}.{ty} {}, [{}];", reg(dst), clean_operand(src))
        }
        Store { dst, value, .. } => {
            let ty = register_type(value, all_instrs);
            let space = if is_local(dst, all_instrs) {
                "local"
            } else {
                "global"
            };
            format!("st.{space}.{ty} {}, {};", mem(dst), reg(value))
        }
        Br {
            cond,
            target_true,
            target_false,
            ..
        } => match (cond, target_false) {
            (Some(c), Some(f)) => format!("@{} bra {};\n    bra {};", reg(c), target_true, f),
            (None, _) => format!("bra {};", target_true),
            _ => "// invalid conditional branch".to_string(),
        },
        Ret { .. } => "ret;".to_string(),
        GetElementPtr {
            dst, base, index, ..
        } => {
            let base_clean = clean_operand(base);
            let dst_clean = clean_operand(dst);
            let offset = format!("{}_offset", dst_clean);
            let calc_offset = format!("mul.lo.s32 %{}, %{}, 4;", offset, clean_operand(index));
            let calc_ptr = format!("add.s32 %{}, %{}, %{};", dst_clean, base_clean, offset);
            format!("{calc_offset}\n    {calc_ptr}")
        }
        Phi { dst, incoming, .. } => {
            let incoming_str = incoming
                .iter()
                .map(|(label, val)| format!("[{}, {}]", reg(val), label))
                .collect::<Vec<_>>()
                .join(", ");
            format!("// phi node {} <- {}", reg(dst), incoming_str)
        }
        Alloca { .. } => String::new(),
        Unhandled { text, .. } => format!("// unhandled: {}", text),
    }
}

use anyhow::Result;
use llvm_ir::Module;
use llvm_parser::parse_llvm_ir_from_str;

pub fn compile_llvm_to_ptx(ir_code: &str) -> Result<String> {
    let module: Module = parse_llvm_ir_from_str(ir_code)?;
    let mut all_instrs: Vec<(String, Vec<Instruction>)> = vec![];

    for func in &module.functions {
        let blocks = llvm_parser::lower(func)?;
        all_instrs.extend(blocks);
    }

    let kernel_name = module
        .functions
        .iter()
        .next()
        .map(|f| f.name.clone())
        .unwrap_or_else(|| "unknown_kernel".into());

    let ptx_lines = lower_function(&kernel_name, &all_instrs, "sm_75");
    Ok(ptx_lines.join("\n"))
}
