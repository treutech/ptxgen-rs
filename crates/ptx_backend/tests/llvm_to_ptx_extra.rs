// Copyright 2025 Raul Estrada <restrada@treutech.io>
// SPDX-License-Identifier: Apache-2.0
//
// This file is part of the PTXGEN-RS project by Treu Technologies.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use llvm_parser::parse_module::parse_module;
use ir_model::Instruction;

fn run_test(filename: &str) -> String {
    use llvm_parser::convert::lower;
    use ptx_backend::lower_function;

    let path = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("examples")
        .join(filename);
    let module = parse_module(&path).expect("Failed to parse module");

    let mut output = String::new();

    for func in module.functions {
        let all_instrs: Vec<(String, Vec<Instruction>)> = func
            .basic_blocks
            .into_iter()
            .map(|block| {
                let instrs = block.instrs.iter().map(lower).collect::<Vec<_>>();
                (block.name.to_string(), instrs)
            })
            .collect();
        
        for line in lower_function(&func.name, &all_instrs, "sm_75") {
            output.push_str(&format!("{}\n", line));
        }
    }

    output
}

#[test]
fn test_ret_only() {
    insta::assert_snapshot!("ret_only", run_test("ret_only.ll"));
}

#[test]
fn test_br_uncond() {
    insta::assert_snapshot!("br_uncond", run_test("br_uncond.ll"));
}

#[test]
fn test_br_cond() {
    insta::assert_snapshot!("br_cond", run_test("br_cond.ll"));
}

#[test]
fn test_phi_basic() {
    insta::assert_snapshot!("phi_basic", run_test("phi_basic.ll"));
}

#[test]
fn test_conflict_type() {
    insta::assert_snapshot!("conflict_type", run_test("conflict_type.ll"));
}
