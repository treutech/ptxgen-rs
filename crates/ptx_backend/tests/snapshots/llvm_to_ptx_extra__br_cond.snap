---
source: crates/ptx_backend/tests/llvm_to_ptx_extra.rs
expression: "run_test(\"br_cond.ll\")"
---
// Function: br_cond
.version 7.0
.target sm_75
.address_size 64

.entry br_cond {
.reg .s32 %out, %x, %y;

    ret;
}
